url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockInnerMultiplication.java

```diff
@@ -0,0 +1,562 @@
+/*
+ * Copyright (c) 2009-2013, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.block;
+
+/**
+ * <p>
+ * Matrix multiplication for the inner row major blocks, typically inside of a {@link org.ejml.data.BlockMatrix64F}.
+ * </p>
+ *
+ * <p>
+ * This code was auto generated by {@link GeneratorBlockInnerMultiplication} and should not be modified directly.
+ * </p>
+ *
+ * @author Peter Abeles
+ */
+public class BlockInnerMultiplication {
+
+    /**
+     * <p>
+     * Performs the follow operation on individual inner blocks:<br>
+     * <br>
+     * C = C + A * B
+     * </p>
+     */
+    public static void blockMultPlus( final double[] dataA, final double []dataB, final double []dataC,
+                                     int indexA, int indexB, int indexC,
+                                     final int heightA, final int widthA, final int widthC) {
+//        for( int i = 0; i < heightA; i++ ) {
+//            for( int k = 0; k < widthA; k++ ) {
+//                for( int j = 0; j < widthC; j++ ) {
+//                    dataC[ i*widthC + j + indexC ] += dataA[i*widthA + k + indexA] * dataB[k*widthC + j + indexB];
+//                }
+//            }
+//        }
+
+        int a = indexA;
+        int rowC = indexC;
+        for( int i = 0; i < heightA; i++ , rowC += widthC ) {
+            int b = indexB;
+
+            final int endC = rowC + widthC;
+            final int endA = a + widthA;
+            while( a != endA ) {//for( int k = 0; k < widthA; k++ ) {
+                double valA = dataA[a++];
+
+                int c = rowC;
+
+                while( c != endC  ) {//for( int j = 0; j < widthC; j++ ) {
+                    dataC[ c++ ] += valA * dataB[ b++ ];
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 52:68, 153:169.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.